tsinfer/formats.py

```diff
@@ -2480,6 +2480,127 @@ def insert_proxy_samples(
+        old ancestors which are very long, as this can adversely affect performance.
+        Older ancestors should generally be shorter than younger ancestors, so
+        truncating the lengths of older ancestors has negligible effect on inference
+        accuracy.
+
+        :param float lower_time_bound: Defines the lower bound of the closed interval
+            where we search for a truncation value for ancestors older than
+            ``upper_time_bound''.
+        :param float upper_time_bound: The length of ancestors with a time greater
+            than or equal to than this bound will be truncated based on the value
+            of the longest ancestor existing between ``lower_time_bound'' and
+            ``upper_time_bound'' (closed interval), multiplied by
+            ``length_multiplier''. Please ensure that the time value
+            provided matches the units used in the :class:`.AncestorData' file, i.e.
+            if your ancestors do not have site times specified, ``upper_time_bound''
+            should be between 0 and 1.
+        :param float length_multiplier: A multiplier for the length of the longest
+            ancestor between ``lower_time_bound'' and ``uppper_time_bound'', i.e.
+            if the longest ancestor in the specified range is 1 megabase, a
+            ``length_multiplier'' of 2 creates a maximum length of 2
+            megabases.
+        :param \\**kwargs: Further arguments passed to the :func:`AncestorData.copy`
+            when creating the new :class:`AncestorData` instance which will be returned.
+
+        :return: A new :class:`.AncestorData` object.
+        :rtype: AncestorData
+        """
+        self._check_finalised()
+        if upper_time_bound <= 0 or lower_time_bound <= 0:
+            raise ValueError("Time bounds cannot be negative")
+        if length_multiplier <= 0:
+            raise ValueError("Length multiplier cannot be negative")
+        if upper_time_bound < lower_time_bound:
+            raise ValueError("Upper bound must be >= lower bound")
+        sites = self.sites_position[:]
+        start = self.ancestors_start[:]
+        end = self.ancestors_end[:]
+        time = self.ancestors_time[:]
+        focal_sites = self.ancestors_focal_sites[:]
+        haplotypes = self.ancestors_haplotype[:]
+        if upper_time_bound > np.max(time) or lower_time_bound > np.max(time):
+            raise ValueError("Time bounds cannot be greater than older ancestor")
+
+        truncated = self.copy(**kwargs)
+        anc_in_bound = np.logical_and(
+            self.ancestors_time[:] >= lower_time_bound,
+            self.ancestors_time[:] <= upper_time_bound,
+        )
+        if np.sum(anc_in_bound) == 0:
+            raise ValueError("No ancestors in time bound")
+        max_length = length_multiplier * np.max(self.ancestors_length[:][anc_in_bound])
+        if max_length > self.sequence_length:
```

Comment:
I'm not sure this is worth raising an error for. It's not going to do any harm if ``max_length`` does happen to be > L, right? It'll just end up being annoying where this error gets tripped in corner cases that don't really matter anyway.